<a href="https://colab.research.google.com/github/NguyenMinhHieu61/DSToolsAndEcosystem/blob/main/Final_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


In [1]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [2]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data.Date, infer_datetime_format=True), y=stock_data.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data.Date, infer_datetime_format=True), y=revenue_data.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [3]:
tesla = yf.Ticker("TSLA")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [4]:
tesla_data = tesla.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [5]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage [https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue](https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ). Save the text of the response as a variable named `html_data`.


In [9]:
url= "https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue"
html_data=requests.get(url).text

Parse the html data using `beautiful_soup`.


In [10]:
soup = BeautifulSoup(html_data,"html5lib")

Using beautiful soup extract the table with `Tesla Quarterly Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column.


In [20]:
import pandas as pd
import requests

# URL của trang chứa bảng dữ liệu
url = "https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue"

# Thiết lập headers với User-Agent để giả lập trình duyệt
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
}

# Sử dụng requests để lấy nội dung của trang
response = requests.get(url, headers=headers)

# Kiểm tra nếu yêu cầu thành công
if response.status_code == 200:
    # Đọc bảng dữ liệu từ nội dung HTML
    tesla_revenue = pd.read_html(response.text, match="Tesla Quarterly Revenue", flavor='bs4')[0]
    # Đổi tên cột
    tesla_revenue = tesla_revenue.rename(columns={
        'Tesla Quarterly Revenue(Millions of US $)': 'Date',
        'Tesla Quarterly Revenue(Millions of US $).1': 'Revenue'
    }, inplace=False)
    # Xóa các ký tự không mong muốn trong cột Revenue
    # tesla_revenue["Revenue"] = tesla_revenue["Revenue"].str.replace(",", "").str.replace("$", "")
    # Hiển thị 5 hàng đầu tiên
    print(tesla_revenue.head())
else:
    print(f"Failed to retrieve data. Status code: {response.status_code}")


  Tesla Quarterly Revenue (Millions of US $)  \
0                                 2024-06-30   
1                                 2024-03-31   
2                                 2023-12-31   
3                                 2023-09-30   
4                                 2023-06-30   

  Tesla Quarterly Revenue (Millions of US $).1  
0                                      $25,500  
1                                      $21,301  
2                                      $25,167  
3                                      $23,350  
4                                      $24,927  


<ipython-input-20-1f659c77aa9c>:18: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tesla_revenue = pd.read_html(response.text, match="Tesla Quarterly Revenue", flavor='bs4')[0]


In [22]:
# tesla_revenue= pd.read_html(url, match="Tesla Quarterly Revenue", flavor='bs4')[0]
tesla_revenue=tesla_revenue.rename(columns = {'Tesla Quarterly Revenue (Millions of US $)': 'Date', 'Tesla Quarterly Revenue (Millions of US $).1': 'Revenue'}, inplace = False)
# tesla_revenue["Revenue"] = tesla_revenue["Revenue"].str.replace(",","").str.replace("$","")
tesla_revenue.head()

Date  Revenue
0  2024-06-30  $25,500
1  2024-03-31  $21,301
2  2023-12-31  $25,167
3  2023-09-30  $23,350
4  2023-06-30  $24,927

<details><summary>Click here if you need help removing the dollar sign and comma</summary>

```
    
If you parsed the HTML table by row and column you can use the replace function on the string
    
    revenue = col[1].text.replace("$", "").replace(",", "")
    
If you use the read_html function you can use the replace function on the string representation of the column

    tesla_revenue["Revenue"] = tesla_revenue["Revenue"].str.replace("$", "").str.replace(",", "")

```

</details>


Remove the rows in the dataframe that are empty strings or are NaN in the Revenue column. Print the entire `tesla_revenue` DataFrame to see if you have any.


In [23]:
tesla_revenue

Date  Revenue
0   2024-06-30  $25,500
1   2024-03-31  $21,301
2   2023-12-31  $25,167
3   2023-09-30  $23,350
4   2023-06-30  $24,927
..         ...      ...
56  2010-06-30      $28
57  2010-03-31      $21
58  2009-12-31      NaN
59  2009-09-30      $46
60  2009-06-30      $27

[61 rows x 2 columns]

<details><summary>Click here if you need help removing the Nan or empty strings</summary>

```
    
If you have NaN in the Revenue column
    
    tesla_revenue.dropna(inplace=True)
    
If you have emtpty string in the Revenue column

    tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

```

</details>


Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [24]:
tesla_revenue.dropna(inplace=True)
tesla_revenue.tail()

Date Revenue
55  2010-09-30     $31
56  2010-06-30     $28
57  2010-03-31     $21
59  2009-09-30     $46
60  2009-06-30     $27

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [25]:
gamestop = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [26]:
gme_data=gamestop.history(period="max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [27]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620128  1.693350  1.603296  1.691666  76216000   
1 2002-02-14 00:00:00-05:00  1.712708  1.716074  1.670627  1.683251  11021600   
2 2002-02-15 00:00:00-05:00  1.683250  1.687458  1.658001  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615921  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage [https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue](https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork-23455606&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ). Save the text of the response as a variable named `html_data`.


In [28]:

html_data=requests.get(url).text

Parse the html data using `beautiful_soup`.


In [29]:
soup = BeautifulSoup(html_data,"html5lib")

Using beautiful soup extract the table with `GameStop Quarterly Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


In [36]:
# URL của trang chứa bảng dữ liệu
url = "https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue"

# Thiết lập headers để giả lập trình duyệt
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
}

# Gửi yêu cầu GET đến URL
response = requests.get(url, headers=headers)

# Kiểm tra xem yêu cầu có thành công không
if response.status_code == 200:
    # Đọc bảng dữ liệu từ nội dung HTML
    gme_revenue = pd.read_html(response.text, match="GameStop Quarterly Revenue", flavor='bs4')[0]

    # Đổi tên cột
    gme_revenue = gme_revenue.rename(columns={
        'GameStop Quarterly Revenue(Millions of US $)': 'Date',
        'GameStop Quarterly Revenue(Millions of US $).1': 'Revenue'
    }, inplace=False)

    # Xóa dấu phẩy và ký hiệu đô la trong cột 'Revenue'
    # gme_revenue["Revenue"] = gme_revenue["Revenue"].str.replace(",", "").str.replace("$", "")

    # Hiển thị 5 hàng đầu tiên
    print(gme_revenue.head())
else:
    print(f"Failed to retrieve data. Status code: {response.status_code}")

gme_revenue=gme_revenue.rename(columns = {'GameStop Quarterly Revenue (Millions of US $)': 'Date', 'GameStop Quarterly Revenue (Millions of US $).1': 'Revenue'}, inplace = False)

gme_revenue.head()

  GameStop Quarterly Revenue (Millions of US $)  \
0                                    2024-07-31   
1                                    2024-04-30   
2                                    2024-01-31   
3                                    2023-10-31   
4                                    2023-07-31   

  GameStop Quarterly Revenue (Millions of US $).1  
0                                            $798  
1                                            $882  
2                                          $1,794  
3                                          $1,078  
4                                          $1,164  


<ipython-input-36-fb283bbada65>:15: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  gme_revenue = pd.read_html(response.text, match="GameStop Quarterly Revenue", flavor='bs4')[0]


Date Revenue
0  2024-07-31    $798
1  2024-04-30    $882
2  2024-01-31  $1,794
3  2023-10-31  $1,078
4  2023-07-31  $1,164

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [37]:
gme_revenue.dropna(inplace=True)
gme_revenue.tail()

Date Revenue
58  2010-01-31  $3,524
59  2009-10-31  $1,835
60  2009-07-31  $1,739
61  2009-04-30  $1,981
62  2009-01-31  $3,492

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`


In [42]:
print("Columns in stock_data:", tesla_data.columns)
print("Columns in revenue_data:", tesla_revenue.columns)

Columns in stock_data: Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Dividends',
       'Stock Splits'],
      dtype='object')
Columns in revenue_data: Index(['Date', 'Revenue'], dtype='object')


In [46]:
def clean_numeric_data(df, column):
    """
    Hàm làm sạch dữ liệu cột: loại bỏ ký tự '$' và dấu phẩy ','
    """
    # Chỉ áp dụng str.replace nếu cột là chuỗi hoặc cần chuyển đổi thành chuỗi
    if df[column].dtype != 'object':
        df[column] = df[column].astype(str)  # Chuyển sang chuỗi nếu cần

    # Xóa ký tự '$' và dấu phẩy ','
    df[column] = df[column].str.replace('$', '', regex=False)
    df[column] = df[column].str.replace(',', '', regex=False)

    # Chuyển đổi giá trị thành kiểu số (float), xử lý lỗi nếu có giá trị không thể chuyển đổi
    df[column] = pd.to_numeric(df[column], errors='coerce')
    return df

def make_graph(stock_data, revenue_data, stock):
    # Làm sạch dữ liệu trước khi tạo biểu đồ
    stock_data = clean_numeric_data(stock_data, 'Close')
    revenue_data = clean_numeric_data(revenue_data, 'Revenue')

    fig = make_subplots(
        rows=2, cols=1,
        shared_xaxes=True,
        subplot_titles=("Historical Share Price", "Historical Revenue"),
        vertical_spacing=.3
    )

    # Kiểm tra và thêm dữ liệu vào biểu đồ
    if 'Date' in stock_data.columns and 'Close' in stock_data.columns:
        fig.add_trace(
            go.Scatter(
                x=pd.to_datetime(stock_data['Date'], infer_datetime_format=True),
                y=stock_data['Close'],
                name="Share Price"
            ),
            row=1, col=1
        )
    else:
        print("Error: 'Date' or 'Close' columns not found in stock_data.")

    if 'Date' in revenue_data.columns and 'Revenue' in revenue_data.columns:
        fig.add_trace(
            go.Scatter(
                x=pd.to_datetime(revenue_data['Date'], infer_datetime_format=True),
                y=revenue_data['Revenue'],
                name="Revenue"
            ),
            row=2, col=1
        )
    else:
        print("Error: 'Date' or 'Revenue' columns not found in revenue_data.")

    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(
        showlegend=False,
        height=900,
        title=stock,
        xaxis_rangeslider_visible=True
    )
    fig.show()


# Ví dụ gọi hàm
make_graph(tesla_data, tesla_revenue, 'Tesla Stock Data Graph')

<ipython-input-46-1e2ca64c09e2>:33: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.

/usr/local/lib/python3.10/dist-packages/_plotly_utils/basevalidators.py:105: FutureWarning:

The behavior of DatetimeProperties.to_pydatetime is deprecated, in a future version this will return a Series containing python datetime objects instead of an ndarray. To retain the old behavior, call `np.array` on the result

<ipython-input-46-1e2ca64c09e2>:45: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.

/usr/local/lib/python3.10/dist-packages/_plotly_utils/basevalidators.py:105: F

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`.


In [47]:
make_graph(gme_data, gme_revenue, 'GameStop Stock Data Graph')

<ipython-input-46-1e2ca64c09e2>:33: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.

/usr/local/lib/python3.10/dist-packages/_plotly_utils/basevalidators.py:105: FutureWarning:

The behavior of DatetimeProperties.to_pydatetime is deprecated, in a future version this will return a Series containing python datetime objects instead of an ndarray. To retain the old behavior, call `np.array` on the result

<ipython-input-46-1e2ca64c09e2>:45: UserWarning:

The argument 'infer_datetime_format' is deprecated and will be removed in a future version. A strict version of it is now the default, see https://pandas.pydata.org/pdeps/0004-consistent-to-datetime-parsing.html. You can safely remove this argument.

/usr/local/lib/python3.10/dist-packages/_plotly_utils/basevalidators.py:105: F

<h2>About the Authors:</h2>

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
